context.go

```diff
@@ -157,21 +200,41 @@ func merge[K comparable, V any](a, b map[K][]V) (merged map[K][]V) {
 	return
 }
 
+// encodeOneAddressType encodes the given addressData values and returns the corresponding WAF object and its refs.
+// If the addressData is empty, it returns nil for the WAF object and an empty ref pool.
+// At this point, if the encoder does not timeout, the only error we can get is an error in case the top level object
+// is a nil map, but this  behaviour is expected since either persistent or ephemeral addresses are allowed to be null
+// one at a time. In this case, EncodeAddresses will return nil contrary to Encode which will return a nil wafObject,
+// which is what we need to send to ddwaf_run to signal that the address data is empty.
+func encodeOneAddressType(addressData map[string]any, timer timer.Timer) (*bindings.WafObject, encoder, error) {
+	encoder := newLimitedEncoder(timer)
+	if addressData == nil {
+		return nil, encoder, nil
+	}
+
+	data, _ := encoder.EncodeAddresses(addressData)
+	if timer.Exhausted() {
+		return nil, encoder, errors.ErrTimeout
+	}
+
+	return data, encoder, nil
+}
+
 // run executes the ddwaf_run call with the provided data on this context. The caller is responsible for locking the
 // context appropriately around this call.
-func (context *Context) run(persistentData, ephemeralData *bindings.WafObject, timeout time.Duration) (Result, error) {
+func (context *Context) run(persistentData, ephemeralData *bindings.WafObject, wafTimer timer.Timer, timeBudget time.Duration) (Result, error) {
 	result := new(bindings.WafResult)
 	defer wafLib.WafResultFree(result)
 
-	ret := wafLib.WafRun(context.cContext, persistentData, ephemeralData, result, uint64(timeout/time.Microsecond))
-
-	context.totalRuntimeNs.Add(result.TotalRuntime)
-	res, err := unwrapWafResult(ret, result)
-	if err == errors.ErrTimeout {
-		context.timeoutCount.Inc()
+	if timeBudget == timer.UnlimitedBudget {
+		timeBudget = timer.UnlimitedBudget / 1000 // Value is too big for the WAF
```

Comment:
How so? Are we not wiring an `int64` all the way? (I'm a bit surprised, but this does not necessarily need changing... maybe just clarify this statement?)